pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +302,473 @@ pub mod pallet {
+						who.clone(),
+						new_swap.currency_in,
+						new_swap.currency_out,
+						new_swap.amount_in,
+						T::BalanceRatio::one(),
+					)?;
+					Self::register_swap(&who, investment_id, kind, swap_id);
+
+					Ok(SwapStatus {
+						swapped: T::Balance::zero(),
+						pending: new_swap.amount_in,
+					})
+				}
+				Some(swap_id) => {
+					let swap = T::TokenSwaps::get_order_details(swap_id)
+						.ok_or(Error::<T>::SwapOrderNotFound)?;
+
+					if swap.is_same_direction(&new_swap)? {
+						T::TokenSwaps::update_order(
+							who.clone(),
+							swap_id,
+							swap.amount_in.ensure_add(new_swap.amount_in)?,
+							T::BalanceRatio::one(),
+						)?;
+
+						Ok(SwapStatus {
+							swapped: T::Balance::zero(),
+							pending: new_swap.amount_in,
+						})
+					} else {
+						let inverse_swap = swap;
+						let new_swap_amount_out = Self::amount_used_to_swap(&new_swap)?;
+
+						match inverse_swap.amount_in.cmp(&new_swap_amount_out) {
+							Ordering::Less => {
+								T::TokenSwaps::update_order(
+									who.clone(),
+									swap_id,
+									inverse_swap.amount_in.ensure_sub(new_swap_amount_out)?,
+									T::BalanceRatio::one(),
+								)?;
+
+								Ok(SwapStatus {
+									swapped: new_swap.amount_in,
+									pending: T::Balance::zero(),
+								})
+							}
+							Ordering::Equal => {
+								T::TokenSwaps::cancel_order(swap_id.clone())?;
+								Self::unregister_swap(&who, investment_id, kind, swap_id);
+
+								Ok(SwapStatus {
+									swapped: new_swap.amount_in,
+									pending: T::Balance::zero(),
+								})
+							}
+							Ordering::Greater => {
+								T::TokenSwaps::cancel_order(swap_id.clone())?;
+								Self::unregister_swap(&who, investment_id, kind, swap_id);
+
+								let inverse_swap_amount_out =
+									Self::amount_used_to_swap(&inverse_swap)?;
+
+								let amount_to_swap =
+									new_swap.amount_in.ensure_sub(inverse_swap_amount_out)?;
+
+								let swap_id = T::TokenSwaps::place_order(
+									who.clone(),
+									new_swap.currency_in,
+									new_swap.currency_out,
+									amount_to_swap,
+									T::BalanceRatio::one(),
+								)?;
+								Self::register_swap(&who, investment_id, kind, swap_id);
+
+								Ok(SwapStatus {
+									swapped: inverse_swap_amount_out,
+									pending: amount_to_swap,
+								})
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
+		type Amount = T::Balance;
+		type CurrencyId = T::CurrencyId;
+		type Error = DispatchError;
+		type InvestmentId = T::InvestmentId;
+
+		fn increase_foreign_investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			foreign_amount: T::Balance,
+			foreign_currency: T::CurrencyId,
+			pool_currency: T::CurrencyId,
+		) -> DispatchResult {
+			ForeignInvestmentInfo::<T>::insert(
+				who,
+				investment_id,
+				Info {
+					foreign_currency,
+					pool_currency,
+				},
+			);
```

Comment:
There can be multiple `foreign_currency` variants comming in for one `AccountId-InvestmentId` combination. This would override the existing information. 

* We can rather use an `NMap` of `(AccountId, InvestmentId, CurrencyId)` - Still store `Info` there thouhg
* Error out if the existing storage `foreign_currency` does not match

Reply:
Make sense, good point 👍🏻. Right now, are we using different foreign currencies, or is it something expected in the future?

Reply:
@mustermeiszer is this a new requirement or an expected requirement? From the current FI implementation I do not see support for different foreign currencies. Every state is defined by `(account, investment_id)`

Reply:
If this is something new or expected but not used right now I would leave it for future improvement

Reply:
The problem here is that you are overriding the storage. IIRC `FI` does always check, that the currencies are matching right now, which is not done here.

Reply:
Ok, this is fixed now to emulate the FI behavior